_posts/2025-12-12-pqc-on-cheriot.markdown

```diff
@@ -0,0 +1,162 @@
+
+With these defined, it is possible to build both libraries as CHERIoT shared libraries.
+This motivated a bit of cleanup.
+We have a device interface for entropy sources, but it wasn't implemented on the Sail model (which doesn't have an entropy source).
+It has a way of exposing the fact that entropy is insecure, so that wasn't a problem, it just needed doing, so I refactored all of the insecure entropy-source drivers to use a common base.
+Most encryption algorithms want an API that fills a buffer with entropy.
+It's nice if these don't all need to touch the driver directly, so I created a compartment that provides this API and exposes it.
+Now, both libraries are simply consumers of this API.
+This also makes it easier to add stateful whitening for entropy drivers for hardware entropy sources that don't do the whitening in hardware.
+
+Most CHERIoT stacks are on the order of 1-2 KiBs.
+The PQC algorithms use much more space.
+More, in fact, than we permitted.
+
+The previous limitation was based on the precision of bounds rounding.
+A CHERI capability compresses the bounds representation by taking advantage of the fact that, for a pointer to an allocation, there is a lot of redundancy between the address of the pointer, the address of the end of the allocation (the top), and the address of the start of the allocation (the base).
+The distance from the address to base and top are stored as floating-point values with a shared exponent.
+In practical terms, this means that the larger an allocation is, the more strongly aligned its start and end addresses must be.
+The same restrictions apply for any capability that grants access to less than an entire object.
+
+When you call a function in another compartment, the switcher will truncate the stack capability so that the callee sees only the bit of the stack that you weren't using.
+The top and base of the stack must be 16-byte aligned (as an ABI requirement), but a very large stack may have hardware requirements for greater alignment and so may require a gap between the bottom of the caller's stack and the top of the callee's.
+
+Fortunately, we'd added an instruction precisely for this kind of use case: `CSetBoundsRoundDown`.
+This takes a capability and a length and truncates it to *at most* that length.
+It was a fairly small tweak to the switcher to make it do this, and a much larger amount of time with SMT solvers to convince ourselves that this was a safe thing to do.
+
+This also showed up a bug in our linker's handling of the `CAPALIGN` directive, which rounds a section's base and size up to the required alignment to be representable.
+This was not working for sections that followed an explicit alignment directive.
+Our stacks must be both at least 16-byte aligned *and* representable as capabilities.
+This is now fixed.
+
+So now we support stacks up to almost 64 KiB.
+We could easily increase this limit but 64 KiB ought to be enough for anyone.
```

Comment:
I know this is a bit flippant, but I think you could do with an argument for why this is the case e.g. can usually use dynamic memory or globals as an alternative.